lib/server.js

```diff
@@ -0,0 +1,198 @@
+var Docker = require('dockerode-promise');
+var debug = require('debug')('docker-exec-websocket-server:lib:server');
+var through = require('through');
+var fs = require('fs');
+var ws = require('ws');
+var slugid = require('slugid');
+var _ = require('lodash');
+var path = require('path');
+var assert = require('assert');
+var url = require('url');
+var msgcode = require('./messagecodes.js');
+
+var dockerSocket = '/var/run/docker.sock';
+var stats = fs.statSync(dockerSocket);
+if (!stats.isSocket()) {
+  throw new Error('Are you sure the docker is running?');
+}
+var docker = new Docker({socketPath: dockerSocket});
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer([code]), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty, dbgWrite) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  var exec;
+  container.exec(options).then((_exec) => {
+    exec = _exec;
+    return exec.start(attachoptions);
+  }).then((execStream) => {
+    //handling output
+    var strout = through(function (data) {
+      dbgWrite('out', data);
+      var buf = muxBuffer(msgcode.stdout, new Buffer(data));
+      this.queue(buf);
+    });
+
+    var strerr = through(function (data) {
+      dbgWrite('err', data);
+      var buf = muxBuffer(msgcode.stderr, new Buffer(data));
+      this.queue(buf);
+    });
+
+    exec.modem.demuxStream(execStream, strout, strerr);
+    var strbuf = through();
+    strbuf.pipe(through((data) => {
+      socket.send(data, {binary: true});
```

Comment:
A simple hack could be (I might be wrong here), but I suspect:
```js
const MAX_OUTSTANDING_BYTES = 8 * 1024 * 1024;
let outStandingBytes = 0;
strbuf.pipe(through(function(data) {
  outStandingBytes += data.length;
  socket.send(data, {binary: true}, () => {
    outStandingBytes -= data.length;
  });
  if (outStandingBytes > MAX_OUTSTANDING_BYTES) {
    this.pause();
  } else {
    this.resume();
  }
}));
```

See `through` docs: https://github.com/dominictarr/through

Note, there is also `through2` which offers something similar...

-----

Also do you really need `strbuf` like that... How many `through` instances do we need?
Can't we just do: `var strbuf = through((data) => {` instead of `strbuf.pipe(through((data) => {`